### ISSUE TITLE
Problems when updating items for continuous ingestion

### ISSUE THREAD

**Issue by KlausIllmayer** [SSHOC/sshoc-marketplace-backend#475]
This is based upon the old discussion at https://gitlab.gwdg.de/sshoc/sshoc-marketplace-backend/-/issues/85#note_536637

Scenario: for continuous ingestion we like to call the PUT call and it should merge as explained in the above mentioned discussion

Problem: I'm not able to get to the expected behavior in the case when there is a change of the item on SSHOMP in between - this change gets lost but in my understanding it should be merged

Steps to reproduce:
1. Created a test source as administrator: `POST /api/sources'
2. Created a tool as System importer (aka ingestion pipeline): `POST /api/tools-services` using minimal data (label, description, source and sourceItemId)
3. Approved this tool as administrator: `PUT /api/tools-services/{persistent-id}` and removing source and sourceItemId (as described in the mentioned disucssion)
4. Changed the approved tool as administrator: `PUT /api/tools-services/{persistent-id}` and added a value in accessibleAt which did not appear by ingest pipeline
5. Update the tool as System importer: `PUT /api/tools-services/{persistent-id}` and changed the description but still no value in accessibleAt

As a result of step 5 I would expect that for the new ingested version of the tool the description is changed (based on the change of the System importer) and that the accessibleAt value stays (based on the change on the SSHOMP by the administrator). But when looking on the return values the accessibleAt value is not there.

@tparkola is there a chance that you look into this behavior and into the mentioned old discussion so that you may able to tell me if I'm doing something wrong or if my expectation is wonrg or if there is a bug in the described update mechanism?

**Comment by tparkola** [comment]
Can you check what is the payload for in the last PUT call (5.)? From what I see - if there is no information on the accessibleAt in this request's payload then the tool is updated to have no such values. This is most probably why you get what you described.

**Comment by tparkola** [comment]
@KlausIllmayer I added `PATCH` method to the `/api/{category}/{persistent-id}` endpoint. This should work as described in the discussion mentioned by you earlier (https://gitlab.gwdg.de/sshoc/sshoc-marketplace-backend/-/issues/85#note_536637) , i.e. when moderating you should not provide `source` and `sourceItemId`, but when ingesting then you should provide `source` and `sourceItemId` (as this is the way the ingest is identified). So now, please use/test `PATCH` in the ingestion pipelines and do not use `PUT`. A comment related to `PUT`: it was doing some comparison of three versions in specific scenarios, as mentioned by you, but the comparison was not proper and cause problems that you indicated. So I removed this behaviour from `PUT`. This way `PUT `does what is should - update the resource provided in the request. Now `PATCH `should be used in ingestion pipelines.
**IMPORTANT**: Because these changes are quite extensive, please also make sure to do testing of other parts of the MP, including user interface, moderation, etc.

**Comment by KlausIllmayer** [comment]
I tested the PATCH endpoint by using the scenario described in the first post of this issue. But instead of using the PUT in step 5 I now used the PATCH. But I still get a result from this patch call, that does ignore the changes done on the marketplace (in this case, adding an accessibleAt-value which is my scenario not present in the source). Maybe I misunderstand something - can you try to reproduce my scenario to see, if you have the same effect?

**Comment by KlausIllmayer** [comment]
Okay, I need to take back my last comment - I tested it again and now it works. Seems that last week I did something wrong, sorry.

But when testing it now I found another issue:

I take the same approach as in my very first approach which now works. But for step 5 (by ingest) I additionally added an accessibleAt-value which is different to the one that was applied in step 4 (by administrator). I expected that this will raise the "conflict-at-source" scenario, which means, that the property `{"conflict-at-source": "true"}` is added to the response of the PATCH call. But this was not the case. @tparkola Could it be, that this behaviour is missing for the PATCH call? If so, would it be possible to add it again (it was already in the old PUT-call implemented).

Side note: it looks like, that it also works with the current approve-workflow, where we don't remove source and sourceItemId and instead use the endpoint `PUT /api/{category}/{persistentId}/revert` - we need still to test this intensively, as it also means that in future edits by contributors and moderators the source and sourceItemId stays on the item. @tparkola what do you think about this approach with the revert-endpoint and do you have a clue why it works even without not removing source and sourceItemId?

**Comment by tparkola** [comment]
I had a look at the revert - it looks like it has never taken into consideration information about `source` or `sourceItemId` as it was introduced based on https://github.com/SSHOC/sshoc-marketplace-backend/issues/435. Does that explain everything?

**Comment by KlausIllmayer** [comment]
Thanks for your answer. It explains it a little bit, but the question itself is a bit different. I reformulate it:

* Do you think, that using the revert-endpoint is still a good way to approve items? Or will this have a negative effect on the implemented PATCH mechanism? In my understanding it could mean, that by approving items with revert (thus not removing source and sourceItemId - and you explicit told us in this [comment ](https://github.com/SSHOC/sshoc-marketplace-backend/issues/475#issuecomment-2457422186) that we should give away source and sourceItemId when moderating/approving) the PATCH will give wrong results as it can't differ between the version of the item that comes from the source and the versions that were updated on the MP.

And there was also a second question in my last comment, about missing the conflict-at-source-property. Could you kindly elaborate also on this question?

**Comment by laureD19** [comment]
PATCH endpoint tested on stage, following approach described by Klaus above.
When `source` and `sourceItemID` are removed before approving the item but kept by the System Importer in PATCH, the result of the PATCH works as expected: change from source taken (change description field) and curation done in MP (adding accessibleAt) kept.

But I agree with you @KlausIllmayer that we should revise our whole moderation workflow to clarify:
1. how approving ingested items making sure we always remove source and sourceItemIDs? This means that we won't see source affiliation on frontend for all accepted items anymore for ex.
2. `source` and `sourceItemID` should be somehow made mandatory for ingested items, if such a thing is possible
3. if `conflict-at-source` doesn't work as expected, we could use the `diff` method to review the items with 'ingested' status, the same way we use the diff method to display items to moderate. Would that be a possible approach for frontend?
4. maybe good to revise and agree on this overview of transition for items and versions, as Michał prepared it back in... 2020: https://docs.google.com/document/d/14JI_vPtwng-m0TqD-v-fg1q_RfU-t8skZBgmVeA9T2s/edit?tab=t.0

**Comment by KlausIllmayer** [comment]
The use case where the PATCH in my opinion should raise a `conflict-at-source` but does not work:

* Get Bearer token authorization of a user that has the role "Administrator" and one that has the role "System importer"
* Create a test source as administrator: `POST /api/sources' (add Bearer token authorization of administrator)
```
curl -X 'POST' \
  'http://localhost/api/sources' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "label": "Test Source",
  "url": "https://test.acdh.oeaw.ac.at",
  "urlTemplate": "https://test.acdh.oeaw.ac.at/id/{source-item-id}"
}'
```
(returns as example source.id: 2503)
* Create a tool as System importer (aka ingestion pipeline) to this new created source: `POST /api/tools-services` using minimal data (label, description, source and sourceItemId) (add Bearer token authorization of System importer; replace source.id with the one that was created in the last point)
```
curl -X 'POST' \
  'http://localhost/api/tools-services?draft=false' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "label": "test tool conflict-at-source",
  "description": "try to update a tool and raise conflict-at-source",
  "source": {
    "id": 2503
  },
  "sourceItemId": "item-1"
}'
```
(returns as example persistentId: uZhaT3)
* Approve this tool as administrator: `PUT /api/tools-services/{persistent-id}` and remove source and sourceItemId as described in the mentioned discussion (add Bearer token authorization of Administrator; take the return payload of the last point)
```
curl -X 'PUT' \
  'http://localhost/api/tools-services/uZhaT3?draft=false&approved=true' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": 79404,
  "category": "tool-or-service",
  "label": "test tool conflict-at-source",
  "persistentId": "uZhaT3",
  "lastInfoUpdate": "2025-06-25T13:20:02+0000",
  "status": "ingested",
  "informationContributor": {
    "id": 4,
    "username": "System importer",
    "displayName": "System importer",
    "status": "enabled",
    "registrationDate": "2021-05-28T15:28:06+0000",
    "role": "system-contributor",
    "config": true
  },
  "description": "try to update a tool and raise conflict-at-source",
  "contributors": [],
  "properties": [],
  "externalIds": [],
  "accessibleAt": [],
  "relatedItems": [],
  "media": []
}'
```
* Change the approved tool as administrator (means: a user on the marketplace did a change): `PUT /api/tools-services/{persistent-id}` and add a value in accessibleAt which did not appear in point 2
```
curl -X 'PUT' \
  'http://localhost/api/tools-services/uZhaT3?draft=false&approved=true' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": 79404,
  "category": "tool-or-service",
  "label": "test tool conflict-at-source",
  "persistentId": "uZhaT3",
  "lastInfoUpdate": "2025-06-25T13:41:55+0000",
  "status": "approved",
  "informationContributor": {
    "id": 1,
    "username": "Administrator",
    "displayName": "Administrator",
    "status": "enabled",
    "registrationDate": "2021-05-28T15:28:06+0000",
    "role": "administrator",
    "email": "administrator@example.com",
    "config": true
  },
  "description": "try to update a tool and raise conflict-at-source",
  "contributors": [],
  "properties": [],
  "externalIds": [],
  "accessibleAt": [
    "https://ingest.acdh.oeaw.ac.at/item1"
  ],
  "relatedItems": [],
  "media": []
}'
```
* Update the tool as System importer: `PATCH /api/tools-services/{persistent-id}` and give a different url in accessibleAt:
```
curl -X 'PATCH' \
  'http://localhost/api/tools-services/uZhaT3?draft=false&approved=true' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "label": "test tool conflict-at-source",
  "description": "try to update a tool and raise conflict-at-source",
  "source": {
    "id": 2503
  },
  "sourceItemId": "item-1",
  "accessibleAt": [
    "https://somewhereelse.acdh.oeaw.ac.at/item1"
  ]
}'
```
This returns:
```
{
  "id": 79407,
  "category": "tool-or-service",
  "label": "test tool conflict-at-source",
  "persistentId": "uZhaT3",
  "lastInfoUpdate": "2025-06-25T13:47:41+0000",
  "status": "ingested",
  "informationContributor": {
    "id": 4,
    "username": "System importer",
    "displayName": "System importer",
    "status": "enabled",
    "registrationDate": "2021-05-28T15:28:06+0000",
    "role": "system-contributor",
    "config": true
  },
  "description": "try to update a tool and raise conflict-at-source",
  "contributors": [],
  "properties": [],
  "externalIds": [],
  "accessibleAt": [
    "https://somewhereelse.acdh.oeaw.ac.at/item1"
  ],
  "source": {
    "id": 2503,
    "label": "Test Source",
    "url": "https://test.acdh.oeaw.ac.at",
    "urlTemplate": "https://test.acdh.oeaw.ac.at/id/{source-item-id}"
  },
  "sourceItemId": "item-1",
  "relatedItems": [],
  "media": []
}
```
In my opinion, it should have additionally added the property `conflict-at-source` as there is conflict between the change in marketplace (adding the accessibleAt pointing to https://ingest.acdh.oeaw.ac.at/item1) and the one by the ingest (having an accessibleAt pointing to https://somewhereelse.acdh.oeaw.ac.at/item1). Instead it proposes to overwrite the change of marketplace user with the accessibleAt of the ingest (= System importer).

@pWoz Maybe I have some wrong assumptions here or do something wrong. Can you confirm my approach and also that here the conflict-at-source should be there? I also tried it without deleting source and sourceItemId (at point 2) but it does also not add the `conflict-at-source`.

**Comment by KlausIllmayer** [comment]
We also can try to use for the second step (the approve) the revert-endpoint (that is what currently happens, if an item is approved) to see, if this changes something. But before I like to see if the assumptions on my side are correct. Afterwards we can discuss how the revert-endpoint suits into this picture.

**Comment by KlausIllmayer** [comment]
@pWoz There is still the issue with the missing property `conflict-at-source` if there are changes both on side of the source and on side of the marketplace. I've elaborated it in this issue including instructions, two comments above, see https://github.com/SSHOC/sshoc-marketplace-backend/issues/475#issuecomment-3004891677

I would need an answer so that I can further evaluate the POST mechanism. The question is: when proceeding in the way as shown in the mentioned comment, I would expect that the property `conflict-at-source` will be added by the marketplace backend. It looks like as if marketplace backend does not add this property. Is it because I use a wrong approach or is it because there is a bug in marketplace backend?

I'm categorizing it as a bug, because I remember that I already saw the `conflict-at-source`-property when doing such merges. But that was before we switched to the POST mechanism. Can you have a look into the source code, if and when this property will be added? This could help in understanding, if this feature is not working as expected. You can also give an estimation, how long it would take to implement this feature.

**Comment by pWoz** [comment]
I introduced the following changes in the code: 
- I modified `ItemsConflictComparator` so it should detect conflicts correctly now;
- I wrote many tests for `ItemsConflictComparator` in here `ItemsConflictComparatorTest`. 

@KlausIllmayer could you have a look on these tests and verify if logic implemented there is correct? You can focus on `when` part and expected test result, which is always true or false indicating that there is conflict or not.

I also went thru the same procedure as Klaus and it seem to be working fine now (there is `conflict-at-source` detected).

Remarks: 
- there is list of all compared properties in here: ItemsConflictComparator.isConflict (one method per one property), can I assume that there is everything that should be compared? (not all fields from `Item` are taken into consideration during conflicts check);
- this mechanism is case sensitive. For example these two values for `accessbleAt`: "sample" and "SAMPLE", will trigger `conflict-at-source`. Is it correct?
-this mechanism is quite restrictive. For example order of the elements in the `accessbleAt` list matters. Is it correct?

**Comment by KlausIllmayer** [comment]
Thanks @pWoz - I will have a look next week

**Comment by KlausIllmayer** [comment]
Looks quite fine to me, the only thing that comes into my mind is in the method `shouldRecognizeItemsAsConflictedBecauseOfMedia_1()` when you use `UUID.randomUUID()` for the new media that should differ, which could in theory have the same UUID as the one used in `prepareItemTemplate()` - but I know that this would be very unlikely.

Regarding your remarks:
* Thanks for the list. ~~I miss digitalObjects.dateLastUpdated and digitalObjects.dateCreated - is there a reason that you didn't add them?~~ [comment: was there but I missed it] And I'm unsure about the workflow steps: there is the composedOf list where all of the contained steps are listed, should this also be compared in your opinion?
* It is correct, that the mechanism should be case sensitive.
* Based on your example, the mechanism can stay restrictive. We will need to find out, if it is maybe to restrictive. We will need some feedback by curators if issues like different order of items in a collection will appear often. I guess that is something where it depends if SSHOMP takes the order of items in a list as pushed by the ingest pipeline. And it also depends if the ingest pipeline (e.g. DACE) also stays with the order. If we see that this is not the case, it could create to much noise, we would then need to rethink the restrictive approach.

**Comment by pWoz** [comment]
I modified generation of UUID in `shouldRecognizeItemsAsConflictedBecauseOfMedia_1()`.

Regarding `composedOf`: As far as I understand the source code, there is no way to modify this value while PATCH'ing workflows and/or workflow steps (it is in `WorkflowController` class). All possible values that maybe modified as part of patching process are in `StepCore` and `WorkflowCore`, and there is no `composedOf`. So, in my opinion, there is no point to compare it. Otherwise we would have to significantly modify patching process.

**Comment by mkrzmr** [comment]
> And I'm unsure about the workflow steps: there is the composedOf list where all of the contained steps are listed, should this also be compared in your opinion?

Considering the planed changes to the workflow entry https://github.com/SSHOC/sshoc-marketplace-backend/issues/470 and because workflows are not (yet) regularly ingested from a source, the current approach is fine in my opinion

**Comment by KlausIllmayer** [comment]
Changes were merged into stage and develop and can be tested there

**Comment by mkrzmr** [comment]
Konrad wrote up the status of CI here : https://docs.google.com/document/d/1qbDyAUu97-zlIo9idQ5dMuEtpcANGTtcemS1jJZqKOo/edit?usp=sharing

**Comment by mkrzmr** [comment]
Here is my perspective on CI: The current approach compares Source to SSHOMP. If there is a difference, it ingests and raises the conflict-at-source flag. 

Expected behaviour: CI compares Source(now) to Source(at time of last ingest). If there is a difference, it ingest the item to SSHOMP, raising the conflict-at-source flag. 

This will likely require more changes, but will be way to signal that the source has made changes to an item that SSHOMP has not seen yet.

**Comment by KlausIllmayer** [comment]
After further investigation and looking into the mentioned conflict-at-source-flag issue, I discovered another scenario which raises problems regarding the continuous ingestion:

1. A system-moderator like an ingest-tool creates a new tool: `POST api/tools-services` with the test data `{"label": "Test PATCH", "description": "Ingested data"}`
2. A moderator approves this ingested item and it shows up to the public: `PUT api/tools-services/{persistentId}/versions/{versionId}` (taking the values from step 1)
3. The moderator then changes the item in the frontend and adds a new property like a discipline
4. Now the continuous ingest is started and the system-moderator patches the item, which does not have any changes at the source: `PATCH api/tools-services/{persistent-id}` with the data from step 1 as there were no changes: `{"label": "Test PATCH", "description": "Ingested data"}`
5. The result is that the item did not change, it still has the additional added property, which is the expected behaviour, and a moderator approves the item.
6. Now the moderator does change the description on the SSHOMP: `PATCH api/tools-services` with the edited data `{"label": "Test PATCH", "description": "Ingested data and edited"}`
7. Again, the continuous ingest is started and the system-moderator patches the item, which again did not change at the source: `PATCH api/tools-services/{persistent-id}` with the data from step 1 as there were no changes: `{"label": "Test PATCH", "description": "Ingested data"}`
8. And here the problem is raised as step 7 will overwrite the change from step 6. It is expected, that changes on side of the SSHOMP are not to be forgotten, but step 7 will set the description to `Ingested data`, instead it should stay on `Ingested data and edited`

@pWoz Could you have a look and comment, if the problem in step 8 is a bug or a conceptual problem? I remember that long time ago the idea was, that when an item is again ingested via the continuous ingestion, there will be a check against the last ingested version of this item (the one in step 4) and only if there was a change, it will be taken into account in the PATCH of step 7. As between step 1 and step 7 there is no difference, it should not touch the description and overwrite changes done in the SSHOMP.

**Comment by KlausIllmayer** [comment]
@kpaw-man Can you maybe have a look at my last comment https://github.com/SSHOC/sshoc-marketplace-backend/issues/475#issuecomment-3973614981 and report back, if you see this also as a problem for the DACE continuous ingest (especially my expectations formulated in point 8)?

**Comment by kpaw-man** [comment]
In MarketPlace backend, I can see the ItemsConflictComparator#isConflict function, which checks multiple elements of the items, including the description. Blame points to 18.12.2021, so I would say this is a conceptual problem.

Regarding CI. When such an element occurs during CI execution, meaning a different description is present in the source,
it may indeed appear as a conflict. However, if the source is relatively stable and mostly inserts new objects rather than changing existing ones,  this scenario should not occur all the time. I would say it depends on the maturity of the source itself.

In short - in the CI context, when objects are stable in the source and only new ones are being added, the problem should not occur.

**Comment by KlausIllmayer** [comment]
But isn't it the case, that for some of our sources there are changes on items happening? I wonder if this is maybe the case for DARIAH-CAMPUS. You didn't have a closer look, if some of the items changed at source? @mkrzmr Do you know by chance, if there are updates of items on side of DARIAH-CAMPUS source?
